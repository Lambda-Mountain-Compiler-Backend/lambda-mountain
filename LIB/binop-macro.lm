
macro ('declare-binop (op-alias op-code op-type op-rtype reg-suffix rreg-suffix reg-type)) (
   op-alias := λ: Blob(: l op-type+Constant)(: r op-type+Constant). (: (
      (:program(
         '\tmov\s$_s (:expression l) ',\s%R13_l reg-suffix '\n_s
         '\tmov\s$_s (:expression r) ',\s%R14_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);
   op-alias := λ: Blob(: l op-type+Constant)(: r op-type+LocalVariable). (: (
      (:program(
         '\tmov\s$_s (:expression l) ',\s%R13_l reg-suffix '\n_s
         '\tmov\s_s (:expression r) '\[%RBP\],\s%R14_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);
   op-alias := λ: Blob(: l op-type+Constant)(: r op-type+GlobalVariable). (: (
      (:program(
         '\tmov\s$_s (:expression l) ',\s%R13_l reg-suffix '\n_s
         '\tmov\s$_s (:expression r) ',\s%R14\n_s
         '\tmov\s0\[%R14\],\s%R14_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);
   op-alias := λ: Blob(: l op-type+Constant)(: r op-type+reg-type). (: (
      (:program(
         (:program r)
         '\tmov\s%_s (:expression r) ',\s%R14_l reg-suffix '\n_s
         '\tmov\s$_s (:expression l) ',\s%R13_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);

   op-alias := λ: Blob(: l op-type+LocalVariable)(: r op-type+Constant). (: (
      (:program(
         '\tmov\s_s (:expression l) '\[%RBP\],\s%R13_l reg-suffix '\n_s
         '\tmov\s$_s (:expression r) ',\s%R14_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);
   op-alias := λ: Blob(: l op-type+LocalVariable)(: r op-type+LocalVariable). (: (
      (:program(
         '\tmov\s_s (:expression l) '\[%RBP\],\s%R13_l reg-suffix '\n_s
         '\tmov\s_s (:expression r) '\[%RBP\],\s%R14_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);
   op-alias := λ: Blob(: l op-type+LocalVariable)(: r op-type+GlobalVariable). (: (
      (:program(
         '\tmov\s_s (:expression l) '\[%RBP\],\s%R13_l reg-suffix '\n_s
         '\tmov\s$_s (:expression r) ',\s%R14\n_s
         '\tmov\s0\[%R14\],\s%R14_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);
   op-alias := λ: Blob(: l op-type+LocalVariable)(: r op-type+reg-type). (: (
      (:program(
         (:program r)
         '\tmov\s%_s (:expression r) ',\s%R14_l reg-suffix '\n_s
         '\tmov\s_s (:expression l) '\[%RBP\],\s%R13_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);

   op-alias := λ: Blob(: l op-type+GlobalVariable)(: r op-type+Constant). (: (
      (:program(
         '\tmov\s$_s (:expression l) ',\s%R13\n_s
         '\tmov\s0\[%R13\],\s%R13_l reg-suffix '\n_s
         '\tmov\s$_s (:expression r) ',\s%R14_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);
   op-alias := λ: Blob(: l op-type+GlobalVariable)(: r op-type+LocalVariable). (: (
      (:program(
         '\tmov\s$_s (:expression l) ',\s%R13\n_s
         '\tmov\s0\[%R13\],\s%R13_l reg-suffix '\n_s
         '\tmov\s_s (:expression r) '\[%RBP\],\s%R14_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);
   op-alias := λ: Blob(: l op-type+GlobalVariable)(: r op-type+GlobalVariable). (: (
      (:program(
         '\tmov\s$_s (:expression l) ',\s%R13\n_s
         '\tmov\s0\[%R13\],\s%R13_l reg-suffix '\n_s
         '\tmov\s$_s (:expression r) ',\s%R14\n_s
         '\tmov\s0\[%R14\],\s%R14_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);
   op-alias := λ: Blob(: l op-type+GlobalVariable)(: r op-type+reg-type). (: (
      (:program(
         (:program r)
         '\tmov\s%_s (:expression r) ',\s%R14_l reg-suffix '\n_s
         '\tmov\s$_s (:expression l) ',\s%R13\n_s
         '\tmov\s0\[%R13\],\s%R13_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);

   op-alias := λ: Blob(: l op-type+reg-type)(: r op-type+Constant). (: (
      (:program(
         (:program l)
         '\tmov\s%_s (:expression l) ',\s%R13_l reg-suffix '\n_s
         '\tmov\s$_s (:expression r) ',\s%R14_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);
   op-alias := λ: Blob(: l op-type+reg-type)(: r op-type+LocalVariable). (: (
      (:program(
         (:program l)
         '\tmov\s%_s (:expression l) ',\s%R13_l reg-suffix '\n_s
         '\tmov\s_s (:expression r) '\[%RBP\],\s%R14_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);
   op-alias := λ: Blob(: l op-type+reg-type)(: r op-type+GlobalVariable). (: (
      (:program(
         (:program l)
         '\tmov\s%_s (:expression l) ',\s%R13_l reg-suffix '\n_s
         '\tmov\s$_s (:expression r) ',\s%R14\n_s
         '\tmov\s0\[%R14\],\s%R14_l reg-suffix '\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);
   op-alias := λ: Blob(: l op-type+reg-type)(: r op-type+reg-type+Meta<OntoStack>). (: (
      (:program(
         (:program r)
         '\tpush\s%_s (:expression r) '\n_s
         (:program l)
         '\tmov\s%_s (:expression l) ',\s%R13_l reg-suffix '\n_s
         '\tpopq\s%R14\n_s
         op-code
      ))
      (:frame( (:frame l) (:frame r) ))
      (:unframe( (:unframe l) (:unframe r) ))
      (:expression( 'R13_l rreg-suffix ))
   ) op-rtype);
);

