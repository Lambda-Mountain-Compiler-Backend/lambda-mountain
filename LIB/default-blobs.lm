
:program := λ: Blob(: x x). (: () Nil);
:frame := λ: Blob(: x x). (: () Nil);
:unframe := λ: Blob(: x x). (: () Nil);
:text := λ: Blob(: x x). (: () Nil);
:data := λ: Blob(: x x). (: () Nil);
:del := λ: Blob(: x x). (: () Nil);
