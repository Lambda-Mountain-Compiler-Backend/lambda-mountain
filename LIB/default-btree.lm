
type BTree<k,v> (BTreeBranch( U64 , k , v , BTree<k,v>[] , BTree<k,v>[] )) | BTreeEOF;

.lookup := λ(: index BTree<k,v>)(: k k)(: default v). (: (tail(
   (let th (hash k))
   (let continue True_u8)
   (while (==( continue True_u8 )) (
      (let ih (.5 (as index BTree<k,v>)))
      (if (==( th ih )) (
         (if (is( k (.4 (as index BTree<k,v>)) )) (tail(
            (set default (.3 (as index BTree<k,v>)))
            (set continue False_u8)
         )) ())
      ) ())
      (if (==( continue True_u8 )) (
         (if (<( ih th )) (
            (if (==( (as (.1 (as index BTree<k,v>)) U64) 0_u64 )) (
               (set continue False_u8)
            ) (
               (set index (maybe-deref(.1 (as index BTree<k,v>))))
            ))
         ) (
            (if (==( (as (.2 (as index BTree<k,v>)) U64) 0_u64 )) (
               (set continue False_u8)
            ) (
               (set index (maybe-deref(.2 (as index BTree<k,v>))))
            ))
         ))
      ) ())
   ))
   default
)) v);

.bind := λ(: index BTree<k,v>[])(: k k)(: v v). (: (tail(
   (let ht (hash k))
   (let continue True_u8)
   (let cell (close(BTreeBranch( ht k v (as 0_u64 BTree<k,v>[]) (as 0_u64 BTree<k,v>[]) ))))
   (while (==( continue True_u8 )) (tail(
      (let flat-index (maybe-deref index))
      (if (<( (.5 (as flat-index BTreeBranch)) ht )) (
         (if (==( (as (.1 (as flat-index BTreeBranch)) U64) 0_u64 )) (tail(
            (mov( (as cell U64) R8 ))
            (set[]( (as index Field<Sized<8>,1>[]) 1_u64 R8 ))
            (set continue False_u8)
         )) (
            (set index (.1 (as flat-index BTreeBranch)))
         ))
      ) (
         (if (==( (as (.2 (as flat-index BTreeBranch)) U64) 0_u64 )) (tail(
            (mov( (as cell U64) R8 ))
            (set[]( (as index Field<Sized<8>,1>[]) 2_u64 R8 ))
            (set continue False_u8)
         )) (
            (set index (.2 (as flat-index BTreeBranch)))
         ))
      ))
   )))
)) Nil);
