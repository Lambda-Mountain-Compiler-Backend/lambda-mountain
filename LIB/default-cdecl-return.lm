
cdecl::return := λ: Blob(: l Nil). (: (
   (:frame( (:frame l) ))
   (:unframe( (:unframe l) ))
   (:program( (:program l) (:del l) ))
) Nil);
cdecl::return := λ: Blob(: l Reg8). (: (
   (:frame( (:frame l) ))
   (:unframe( (:unframe l) ))
   (:expression( 'AL_s ))
   (:program(
      (:program l)
      '\tmov\t%_s (:expression l) ',%AL\n_s
      '\tpush\s%RAX\n_s
      (:del l)
      '\tpop\s%RAX\n_s
   ))
) Nil);
cdecl::return := λ: Blob(: l Reg16). (: (
   (:frame( (:frame l) ))
   (:unframe( (:unframe l) ))
   (:expression( 'AX_s ))
   (:program(
      (:program l)
      '\tmov\t%_s (:expression l) ',%AX\n_s
      '\tpush\s%RAX\n_s
      (:del l)
      '\tpop\s%RAX\n_s
   ))
) Nil);
cdecl::return := λ: Blob(: l Reg32). (: (
   (:frame (:frame l))
   (:unframe( (:unframe l) ))
   (:expression 'EAX_s)
   (:program(
      (:program l)
      '\tmov\t%_s (:expression l) ',%EAX\n_s
      '\tpush\s%RAX\n_s
      (:del l)
      '\tpop\s%RAX\n_s
   ))
) Nil);
cdecl::return := λ: Blob(: l Reg64). (: (
   (:frame (:frame l))
   (:unframe( (:unframe l) ))
   (:expression 'RAX_s)
   (:program(
      (:program l)
      '\tmov\t%_s (:expression l) ',%RAX\n_s
      '\tpush\s%RAX\n_s
      (:del l)
      '\tpop\s%RAX\n_s
   ))
) Nil);

cdecl::return := λ: Blob(: l Constant+Literal+Sized<1>). (: (
   (:program(
      (:program l)
      (:del l)
      '\tmov\t$_l (:expression l) ',%AL\n_l
   ))
   (:frame (:frame l))
   (:unframe (:unframe l))
   (:expression 'AL_l)
) Nil);
cdecl::return := λ: Blob(: l Constant+Literal+Sized<2>). (: (
   (:program(
      (:program l)
      (:del l)
      '\tmov\t$_l (:expression l) ',%AX\n_l
   ))
   (:frame (:frame l))
   (:unframe (:unframe l))
   (:expression 'AX_l)
) Nil);
cdecl::return := λ: Blob(: l Constant+Literal+Sized<4>). (: (
   (:program(
      (:program l)
      (:del l)
      '\tmov\t$_l (:expression l) ',%EAX\n_l
   ))
   (:frame (:frame l))
   (:unframe (:unframe l))
   (:expression 'EAX_l)
) Nil);
cdecl::return := λ: Blob(: l Constant+Literal+Sized<8>). (: (
   (:program(
      (:program l)
      (:del l)
      '\tmov\t$_l (:expression l) ',%RAX\n_l
   ))
   (:frame (:frame l))
   (:unframe (:unframe l))
   (:expression 'RAX_l)
) Nil);

cdecl::return := λ: Blob(: src LocalVariable+Sized<size>). (: (
   (:program(
      (:program src)
      (:del src)
      (if (==( (: size L) 1_l )) ( '\tmov\s_l (:expression src) '\[%rbp\],\s%AL\n_l ) ())
      (if (==( (: size L) 2_l )) ( '\tmov\s_l (:expression src) '\[%rbp\],\s%AX\n_l ) ())
      (if (==( (: size L) 4_l )) ( '\tmov\s_l (:expression src) '\[%rbp\],\s%EAX\n_l ) ())
      (if (==( (: size L) 8_l )) ( '\tmov\s_l (:expression src) '\[%rbp\],\s%RAX\n_l ) ())
      (if (>( (: size L) 8_l )) (
         '\tsubq\s$_l (: size L) ',\s%rsp\n_l
         (for-atom (word-offset in (range( 0_l (/( (: size L) 8_l )) ))) (
            '\tpushq\s_l (+( (:expression src) (*( (:expression word-offset) 8_l )) )) '\[%rbp\]\n_l
         ))
         (for-atom (word-offset in (range( 0_l (/( (: size L) 8_l )) ))) (
            '\tpopq\s_l (-( (-( -16_l (*( (:expression word-offset) 8_l )) )) (:expression cdecl::args-size) )) '\[%rbp\]\n_l
         ))
         '\taddq\s$_l (: size L) ',\s%rsp\n_l
      ) ())
   ))
   (:frame (:frame src))
   (:unframe (:unframe src))
) Nil);

cdecl::return := λ: Blob(: src GlobalVariable+Sized<size>). (: (
   (:program(
      (:program src)
      (:del src)
      '\tmov\t$_l (:expression src) '\s,\s%r15\n_l
      (if (==( (: size L) 1_l )) ( '\tmov\s_l (:expression src) '\[%r15\],\s%AL\n_l ) ())
      (if (==( (: size L) 2_l )) ( '\tmov\s_l (:expression src) '\[%r15\],\s%AX\n_l ) ())
      (if (==( (: size L) 4_l )) ( '\tmov\s_l (:expression src) '\[%r15\],\s%EAX\n_l ) ())
      (if (==( (: size L) 8_l )) ( '\tmov\s_l (:expression src) '\[%r15\],\s%RAX\n_l ) ())
      (if (>( (: size L) 8_l )) (
         '\tsubq\s$_l (: size L) ',\s%rsp\n_l
         (for-atom (word-offset in (range( 0_l (/( (: size L) 8_l )) ))) (
            '\tpushq\s_l (*( (:expression word-offset) 8_l )) '\[%r15\]\n_l
         ))
         (for-atom (word-offset in (range( 0_l (/( (: size L) 8_l )) ))) (
            '\tpopq\s_l (-( (-( -16_l (*( (:expression word-offset) 8_l )) )) (:expression cdecl::args-size) )) '\[%rbp\]\n_l
         ))
         '\taddq\s$_l (: size L) ',\s%rsp\n_l
      ) ())
   ))
   (:frame (:frame src))
   (:unframe (:unframe src))
) Nil);

fragment  cdecl::return := λ: DontChain(: src StackVariable+Sized<size>). (: (
   (.program(
      (.program src)
      (.del src)
      (if-eq (.expression size) 1 (\t 'mov \t (.expression src) \[ '%rbp \] , \s '%AL \n))
      (if-eq (.expression size) 2 (\t 'mov \t (.expression src) \[ '%rbp \] , \s '%AX \n))
      (if-eq (.expression size) 4 (\t 'mov \t (.expression src) \[ '%rbp \] , \s '%EAX \n))
      (if-eq (.expression size) 8 (\t 'mov \t (.expression src) \[ '%rbp \] , \s '%RAX \n))
      (if-eq 1 (>( (.expression size) 8 )) (
         \t 'subq \s '$ (.expression size) , \s '%rsp \n
         (for word-offset in (range( 0 (/( (.expression size) 8 )) )) (
            \t 'pushq \s (+( (.expression size) (*( (.expression word-offset) 16 )) )) \[ '%rsp \] \n
         ))
         (for word-offset in (range( 0 (/( (.expression size) 8 )) )) (
            \t 'popq \s (-( (-( '-16 (*( (.expression word-offset) 8 )) )) (.expression cdecl::args-size) )) \[ '%rbp \] \n
         ))
         \t 'addq \s '$ (.expression size) , \s '%rsp \n
      ))
   ))
   (.expression(
      (if-eq (.expression size) 1 'AL)
      (if-eq (.expression size) 2 'AX)
      (if-eq (.expression size) 4 'EAX)
      (if-eq (.expression size) 8 'RAX)
   ))
) Nil);

