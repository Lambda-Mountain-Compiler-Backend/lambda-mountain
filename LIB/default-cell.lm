
type Cell<x> (Cell( x[] ));

cell := λ(: x x). (: (tail(
   (let c (Cell( (close x) )))
   c
)) Cell<x>);

deref := λ(: c Cell<x>). (: (tail(
   (let x (maybe-deref(.1(as c Tag<Cell>))))
   x
)) x);

set[] := λ(: c Cell<x>)(: x x). (: (
   (set[]( (.1(as c Tag<Cell>)) x ))
) Nil);
