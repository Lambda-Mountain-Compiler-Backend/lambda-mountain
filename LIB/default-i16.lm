
meta
   (description (
      'I16\sis\san\sinteger\sof\sthe\srange\s-32768\sto\s32767.
   ))
   (example (print '123_u8))
;
fragment type I16;    size I16 2;    atom suffix I16    _i16;
