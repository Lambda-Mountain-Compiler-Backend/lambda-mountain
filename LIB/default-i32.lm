
meta
   (description (
      'I32\sis\san\sinteger\sof\sthe\srange\s-21474823648\sto\s21474823647.
   ))
   (example (print '123_i32))
;
fragment type I32;    size I32 4;    atom suffix I32    _i32;

#(declare-simple-instruction-alias-binop-32( && 'andl I32 ));
#(declare-simple-instruction-alias-binop-32( || 'orl I32 ));
#(declare-simple-instruction-alias-binop-32( + 'addl I32 ));
#(declare-simple-instruction-alias-binop-32( - 'subl I32 ));
#(declare-simple-instruction-alias-binop-32( ^ 'xorl I32 ));
