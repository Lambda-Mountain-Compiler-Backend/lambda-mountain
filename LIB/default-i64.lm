
meta
   (description (
      'I64\sis\san\sinteger\sof\sthe\srange\s-9223372036854775808\sto\s9223372036854775807.
   ))
   (example (print '123_i64))
;
fragment type I64;    size I64 8;    atom suffix I64    _i64;

#(declare-simple-instruction-alias-binop-64( && 'andq I64 ));
#(declare-simple-instruction-alias-binop-64( || 'orq I64 ));
#(declare-simple-instruction-alias-binop-64( + 'addq I64 ));
#(declare-simple-instruction-alias-binop-64( - 'subq I64 ));
#(declare-simple-instruction-alias-binop-64( ^ 'xorq I64 ));
