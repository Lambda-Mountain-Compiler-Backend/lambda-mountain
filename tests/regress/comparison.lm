
import LIB/default.lm;
#import LIB/dev-u64.lm;

g0 := 0_u64;
g1 := 1_u64;
g2 := 2_u64;

f0 := 位. 0_u64;
f1 := 位. 1_u64;
f2 := 位. 2_u64;

main := 位.(: (
   (let l0 1_u64)
   (let l1 1_u64)
   (let l2 2_u64)

   # Constant, Constant
   (print (==( 1_u64 1_u64 )))
   (print (==( 1_u64 0_u64 )))
   (print '\n_s)
   (print (!=( 1_u64 1_u64 )))
   (print (!=( 1_u64 0_u64 )))
   (print '\n_s)
   (print (<( 1_u64 1_u64 )))
   (print (<( 1_u64 2_u64 )))
   (print '\n_s)
   (print (<=( 1_u64 0_u64 )))
   (print (<=( 1_u64 1_u64 )))
   (print (<=( 1_u64 2_u64 )))
   (print '\n_s)
   (print (>( 1_u64 0_u64 )))
   (print (>( 1_u64 1_u64 )))
   (print '\n_s)
   (print (>=( 1_u64 0_u64 )))
   (print (>=( 1_u64 1_u64 )))
   (print (>=( 1_u64 2_u64 )))
   (print '\n_s)
   (print '\n_s)

   # Constant, Global
   (print (==( 1_u64 g1 )))
   (print (==( 1_u64 g0 )))
   (print '\n_s)
   (print (!=( 1_u64 g1 )))
   (print (!=( 1_u64 g0 )))
   (print '\n_s)
   (print (<( 1_u64 g1 )))
   (print (<( 1_u64 g2 )))
   (print '\n_s)
   (print (<=( 1_u64 g0 )))
   (print (<=( 1_u64 g1 )))
   (print (<=( 1_u64 g2 )))
   (print '\n_s)
   (print (>( 1_u64 g0 )))
   (print (>( 1_u64 g1 )))
   (print '\n_s)
   (print (>=( 1_u64 g0 )))
   (print (>=( 1_u64 g1 )))
   (print (>=( 1_u64 g2 )))
   (print '\n_s)
   (print '\n_s)

   # Constant, Local
   (print (==( 1_u64 l1 )))
   (print (==( 1_u64 l0 )))
   (print '\n_s)
   (print (!=( 1_u64 l1 )))
   (print (!=( 1_u64 l0 )))
   (print '\n_s)
   (print (<( 1_u64 l1 )))
   (print (<( 1_u64 l2 )))
   (print '\n_s)
   (print (<=( 1_u64 l0 )))
   (print (<=( 1_u64 l1 )))
   (print (<=( 1_u64 l2 )))
   (print '\n_s)
   (print (>( 1_u64 l0 )))
   (print (>( 1_u64 l1 )))
   (print '\n_s)
   (print (>=( 1_u64 l0 )))
   (print (>=( 1_u64 l1 )))
   (print (>=( 1_u64 l2 )))
   (print '\n_s)
   (print '\n_s)

   # Constant, Register
   (print (==( 1_u64 (f1()) )))
   (print (==( 1_u64 (f0()) )))
   (print '\n_s)
   (print (!=( 1_u64 (f1()) )))
   (print (!=( 1_u64 (f0()) )))
   (print '\n_s)
   (print (<( 1_u64 (f1()) )))
   (print (<( 1_u64 (f2()) )))
   (print '\n_s)
   (print (<=( 1_u64 (f0()) )))
   (print (<=( 1_u64 (f1()) )))
   (print (<=( 1_u64 (f2()) )))
   (print '\n_s)
   (print (>( 1_u64 (f0()) )))
   (print (>( 1_u64 (f1()) )))
   (print '\n_s)
   (print (>=( 1_u64 (f0()) )))
   (print (>=( 1_u64 (f1()) )))
   (print (>=( 1_u64 (f1()) )))
   (print '\n_s)
   (print '\n_s)

   (malloc( 24_u64 ))
   ()
) Nil)
