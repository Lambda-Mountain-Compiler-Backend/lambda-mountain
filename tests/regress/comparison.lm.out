10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110

10
01
01
011
10
110
